File: Documentation/5.4/Samples/csharp/Raven.Documentation.Samples/Indexes/Boosting.cs
using System.Collections.Generic;
using System.Linq;
using Raven.Client.Documents;
using Raven.Client.Documents.Indexes;
using Raven.Client.Documents.Linq.Indexing;
using Raven.Documentation.Samples.Orders;

namespace Raven.Documentation.Samples.Indexes
{
    #region index_1
    public class Orders_ByCountries_BoostByField : AbstractIndexCreationTask<Order>
    {
        public class IndexEntry
        {
            // Index-field 'ShipToCountry' will be boosted in the map definition below
            public string ShipToCountry { get; set; }
            public string CompanyCountry { get; set; }
        }
        
        public Orders_ByCountries_BoostByField()
        {
            Map = orders => from order in orders
                let company = LoadDocument<Company>(order.Company)
                
                // Note: with current server version,
                // use 'select new' instead of 'select new IndexEntry' for compilation
                select new 
                {
                    // Boost index-field 'ShipToCountry':
                    // * Use method 'Boost', pass a numeric value to boost by 
                    // * Documents that match the query criteria for this field will rank higher
                    ShipToCountry = order.ShipTo.Country.Boost(10), 
                    CompanyCountry = company.Address.Country
                };
        }
    }
    #endregion
    
    #region index_1_js
    public class Orders_ByCountries_BoostByField_JS : AbstractJavaScriptIndexCreationTask
    {
        public Orders_ByCountries_BoostByField_JS()
        {
            Maps = new HashSet<string>()
            {
                @"map('orders', function(order) {
                    let company = load(order.Company, 'Companies')
                    return {
                        ShipToCountry: boost(order.ShipTo.Country, 10),
                        CompanyCountry: company.Address.Country
                    };
                })"
            };
        }
    }
    #endregion
    
    #region index_2
    public class Orders_ByCountries_BoostByIndexEntry : AbstractIndexCreationTask<Order>
    {
        public class IndexEntry
        {
            public string ShipToCountry { get; set; }
            public string CompanyCountry { get; set; }
        }
        
        public Orders_ByCountries_BoostByIndexEntry()
        {
            Map = orders => from order in orders
                let company = LoadDocument<Company>(order.Company)
                
                select new IndexEntry()
                {
                    ShipToCountry = order.ShipTo.Country,
                    CompanyCountry = company.Address.Country
                }
                // Boost the whole index-entry:
                // * Use method 'Boost'
                // * Pass a document-field that will set the boost level dynamically per document indexed.  
                // * The boost level will vary from one document to another based on the value of this field.
               .Boost((float) order.Freight);
        }
    }
    #endregion
    
    #region index_2_js
    public class Orders_ByCountries_BoostByIndexEntry_JS : AbstractJavaScriptIndexCreationTask
    {
        public Orders_ByCountries_BoostByIndexEntry_JS()
        {
            Maps = new HashSet<string>()
            {
                @"map('orders', function(order) {
                    let company = load(order.Company, 'Companies')
                    return boost({
                        ShipToCountry: order.ShipTo.Country,
                        CompanyCountry: company.Address.Country
                    }, order.Freight)
                })"
            };
        }
    }
    #endregion

    public class Boosting
    {
        public async void Queries()
        {
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    #region query_1
                    List<Order> orders = session
                         // Query the index    
                        .Query<Orders_ByCountries_BoostByField.IndexEntry, Orders_ByCountries_BoostByField>()
                        .Where(x => x.ShipToCountry == "Poland" || x.CompanyCountry == "Portugal")
                        .OfType<Order>()
                        .ToList();
                    
                    // Because index-field 'ShipToCountry' was boosted (inside the index definition),
                    // then documents containing 'Poland' in their 'ShipTo.Country' field will get a higher score than
                    // documents containing a company that is located in 'Portugal'.
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region query_2
                    List<Order> orders = await asyncSession
                         // Query the index    
                        .Query<Orders_ByCountries_BoostByField.IndexEntry, Orders_ByCountries_BoostByField>()
                        .Where(x => x.ShipToCountry == "Poland" || x.CompanyCountry == "Portugal")
                        .OfType<Order>()
                        .ToListAsync();
                    
                    // Because index-field 'ShipToCountry' was boosted (inside the index definition),
                    // then documents containing 'Poland' in their 'ShipTo.Country' field will get a higher score than
                    // documents containing a company that is located in 'Portugal'.
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region query_3
                    List<Order> orders = session.Advanced
                         // Query the index 
                        .DocumentQuery<Orders_ByCountries_BoostByField.IndexEntry, Orders_ByCountries_BoostByField>()
                        .WhereEquals(x => x.ShipToCountry, "Poland")
                        .OrElse()
                        .WhereEquals(x => x.CompanyCountry, "Portugal")
                        .OfType<Order>()
                        .ToList();
                    
                    // Because index-field 'ShipToCountry' was boosted (inside the index definition),
                    // then documents containing 'Poland' in their 'ShipTo.Country' field will get a higher score than
                    // documents containing a company that is located in 'Portugal'.
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region query_4
                    List<Order> orders = session
                         // Query the index  
                        .Query<Orders_ByCountries_BoostByIndexEntry.IndexEntry, Orders_ByCountries_BoostByIndexEntry>()
                        .Where(x => x.ShipToCountry == "Poland" || x.CompanyCountry == "Portugal")
                        .OfType<Order>()
                        .ToList();
                    
                    // The resulting score per matching document is affected by the value of the document-field 'Freight'. 
                    // Documents with a higher 'Freight' value will rank higher.
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region query_5
                    List<Order> orders = await asyncSession
                         // Query the index  
                        .Query<Orders_ByCountries_BoostByIndexEntry.IndexEntry, Orders_ByCountries_BoostByIndexEntry>()
                        .Where(x => x.ShipToCountry == "Poland" || x.CompanyCountry == "Portugal")
                        .OfType<Order>()
                        .ToListAsync();
                    
                    // The resulting score per matching document is affected by the value of the document-field 'Freight'. 
                    // Documents with a higher 'Freight' value will rank higher.
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region query_6
                    List<Order> orders = session.Advanced
                         // Query the index  
                        .DocumentQuery<Orders_ByCountries_BoostByIndexEntry.IndexEntry, Orders_ByCountries_BoostByIndexEntry>()
                        .WhereEquals(x => x.ShipToCountry, "Poland")
                        .OrElse()
                        .WhereEquals(x => x.CompanyCountry, "Portugal")
                        .OfType<Order>()
                        .ToList();
                    
                    // The resulting score per matching document is affected by the value of the document-field 'Freight'. 
                    // Documents with a higher 'Freight' value will rank higher.
                    #endregion
                }
            }
        }
    }
}

File: Documentation/6.0/Samples/php/Indexes/Boosting.php
<?php

namespace RavenDB\Samples\Indexes;

use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Indexes\AbstractJavaScriptIndexCreationTask;
use RavenDB\Samples\Infrastructure\Orders\Order;

class Boosting
{
    public function samples(): void
    {
        $store = new DocumentStore();
        try {
            $session = $store->openSession();
            try {
                # region query_1
                $orders = $session
                     // Query the index
                    ->query(Orders_ByCountries_BoostByField_IndexEntry::class, Orders_ByCountries_BoostByField::class)
                    ->whereEquals("ShipToCountry", "Poland")
                    ->orElse()
                    ->whereEquals("CompanyCountry", "Portugal")
                    ->ofType(Order::class)
                    ->toList();

                // Because index-field 'ShipToCountry' was boosted (inside the index definition),
                // then documents containing 'Poland' in their 'ShipTo.Country' field will get a higher score than
                // documents containing a company that is located in 'Portugal'.
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region query_3
                $orders = $session->advanced()
                     // Query the index
                    ->documentQuery(Orders_ByCountries_BoostByField_IndexEntry::class, Orders_ByCountries_BoostByField::class)
                    ->whereEquals("ShipToCountry", "Poland")
                    ->orElse()
                    ->whereEquals("CompanyCountry", "Portugal")
                    ->ofType(Order::class)
                    ->toList();

                // Because index-field 'ShipToCountry' was boosted (inside the index definition),
                // then documents containing 'Poland' in their 'ShipTo.Country' field will get a higher score than
                // documents containing a company that is located in 'Portugal'.
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                    # region query_4
                    $orders = $session
                         // Query the index
                        ->query(Orders_ByCountries_BoostByIndexEntry_IndexEntry::class, Orders_ByCountries_BoostByIndexEntry::class)
                        ->whereEquals("ShipToCountry", "Poland")
                        ->orElse()
                        ->whereEquals("CompanyCountry", "Portugal")
                        ->ofType(Order::class)
                        ->toList();

                    // The resulting score per matching document is affected by the value of the document-field 'Freight'.
                    // Documents with a higher 'Freight' value will rank higher.
                    # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region query_6
                $orders = $session->advanced()
                     // Query the index
                    ->documentQuery(Orders_ByCountries_BoostByIndexEntry_IndexEntry::class, Orders_ByCountries_BoostByIndexEntry::class)
                    ->whereEquals("ShipToCountry", "Poland")
                    ->orElse()
                    ->whereEquals("CompanyCountry", "Portugal")
                    ->ofType(Order::class)
                    ->toList();

                // The resulting score per matching document is affected by the value of the document-field 'Freight'.
                // Documents with a higher 'Freight' value will rank higher.
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}

//# region index_1
class Orders_ByCountries_BoostByField_IndexEntry
{
    // Index-field 'ShipToCountry' will be boosted in the map definition below
    public ?string $shipToCountry = null;
    public ?string $companyCountry = null;

    public function getShipToCountry(): ?string
    {
        return $this->shipToCountry;
    }

    public function setShipToCountry(?string $shipToCountry): void
    {
        $this->shipToCountry = $shipToCountry;
    }

    public function getCompanyCountry(): ?string
    {
        return $this->companyCountry;
    }

    public function setCompanyCountry(?string $companyCountry): void
    {
        $this->companyCountry = $companyCountry;
    }
}

class Orders_ByCountries_BoostByField extends AbstractIndexCreationTask
{
        public function __construct()
        {
            parent::__construct();

            // Boost index-field 'ShipToCountry':
            // * Use method 'Boost', pass a numeric value to boost by
            // * Documents that match the query criteria for this field will rank higher
            $this->map =
                "docs.Orders.Select(order => new { " .
                "   ShipToCountry = order.ShipTo.Country.Boost(10), " .
                "   CompanyCountry = this.LoadDocument(order.Company, \"Companies\").Address.Country " .
                "})";
        }
}
# endregion

# region index_1_js
class Orders_ByCountries_BoostByField_JS extends AbstractJavaScriptIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->setMaps(["map('Orders', function (order) {\n" .
            "   let company = load(order.Company, 'Companies')\n" .
            "   return {\n" .
            "       ShipToCountry: boost(order.ShipTo.Country, 10),\n" .
            "       CompanyCountry: company.Address.Country\n" .
            "   }\n" .
            "})"]);
    }
}
# endregion

# region index_2
class Orders_ByCountries_BoostByIndexEntry_IndexEntry
{
    public ?string $shipToCountry = null;
    public ?string $companyCountry = null;

    public function getShipToCountry(): ?string
    {
        return $this->shipToCountry;
    }

    public function setShipToCountry(?string $shipToCountry): void
    {
        $this->shipToCountry = $shipToCountry;
    }

    public function getCompanyCountry(): ?string
    {
        return $this->companyCountry;
    }

    public function setCompanyCountry(?string $companyCountry): void
    {
        $this->companyCountry = $companyCountry;
    }
}

class Orders_ByCountries_BoostByIndexEntry extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        // Boost the whole index-entry:
        // * Use method 'Boost'
        // * Pass a document-field that will set the boost level dynamically per document indexed.
        // * The boost level will vary from one document to another based on the value of this field.

        $this->map =
            "docs.Orders.Select(order => new { " .
            "   ShipToCountry = order.ShipTo.Country, " .
            "   CompanyCountry = this.LoadDocument(order.Company, \"Companies\").Address.Country " .
            "}.Boost((float) order.Freight))";
    }
}
# endregion

# region index_2_js
class Orders_ByCountries_BoostByIndexEntry_JS extends AbstractJavaScriptIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();


        $this->setMaps(["map('Orders', function (order) {\n" .
            "   let company = load(order.Company, 'Companies')\n" .
            "   return boost({\n" .
            "       ShipToCountry: order.ShipTo.Country,\n" .
            "       CompanyCountry: company.Address.Country\n" .
            "   }, order.Freight)\n" .
            "})"]);
    }
}
# endregion
